src/metabase/api/public.clj

```diff
@@ -531,6 +553,44 @@
      :export-format :api
      :parameters    parameters :qp-runner qp.pivot/run-pivot-query)))
 
+(def ^:private action-execution-throttle
+  "Rate limit at 1 action per second on a per action basis.
+   The goal of rate limiting should be to prevent very obvioius abuse, but it should
+   be relatively lax so we don't annoy legitimate users."
+  (throttle/make-throttler :action-uuid :attempts-threshold 1 :initial-delay-ms 1000 :delay-exponent 1))
+
+#_{:clj-kondo/ignore [:deprecated-var]}
+(api/defendpoint-schema POST "/action/:uuid/execute"
+  "Execute the Action.
+
+   `parameters` should be the mapped dashboard parameters with values.
+   `extra_parameters` should be the extra, user entered parameter values."
+  [uuid :as {{:keys [parameters], :as _body} :body}]
+  {uuid        su/UUIDString
+   parameters  (s/maybe {s/Keyword s/Any})}
+  (let [throttle-message (try
+                           (throttle/check action-execution-throttle uuid)
+                           nil
+                           (catch ExceptionInfo e
+                             (get-in (ex-data e) [:errors :action-uuid])))
+        throttle-time (when throttle-message
+                        (second (re-find #"You must wait ([0-9]+) seconds" throttle-message)))]
```

Comment:
This seems correct to me, but it feels like [`calculate-delay`](https://github.com/metabase/throttle/blob/master/src/throttle/core.clj#L121) should be made public and used directly? @camsaul any thoughts?

Reply:
Or we could make `[make-throttling-ex](make-throttling-ex)` return a map with the `delay-ms` and the message in two values.